webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -26,141 +24,190 @@ interface ScaleZoomData {
-            let max = scalesZoomData.current[scaleKey]?.max;
-            let min = scalesZoomData.current[scaleKey]?.min;
-
-            if (max == null || currentMax > max) max = currentMax;
-            if (min == null || currentMin < min) min = currentMin;
-
-            scalesZoomData.current[scaleKey] = { isZoomed: isZoomed.current, max, min };
-
-            /*
-             * Save the scale info if zoomed in and clear it otherwise.
-             * This info will be used to restore the zoom when remounting
-             * the chart, which can be caused by new series data, chart option
-             * changes, etc.
-             */
-            if (!scalesRef.current) scalesRef.current = {};
-            if (isZoomed.current) {
-              scalesRef.current[scaleKey] = uPlot.scales[scaleKey];
-            } else {
-              delete scalesRef.current[scaleKey];
-            }
-            if (Object.keys(scalesRef.current).length === 0) scalesRef.current = undefined;
-          } ],
-        },
-        scales: scalesRef.current,
-        width: chartDivRef.current.offsetWidth,
-      },
-      options,
-    ) as uPlot.Options;
-
-    const plotChart = new uPlot(optionsExtended, normalizedData, chartDivRef.current);
-
+    if (!chartDivRef.current) return;
+    scalesZoomData.current = {};
+    const data = [ [], [ [] ] ] as unknown as uPlot.AlignedData;
+    if (!chartRef?.current) {
+      chartRef.current = new uPlot(optionsRef.current, data, chartDivRef.current);
+      setChartIsMounted(true);
+    }
     return () => {
-      plotChart.destroy();
+      chartRef?.current?.destroy();
+      chartRef.current = undefined;
+      setChartIsMounted(false);
+    };
+  }, [ ]);
+
+  useEffect(() => {
+    if (!chartDivRef.current) return;
+    let newOptions = getAugmentedOptions(options);
+    if (shouldUpdate(optionsRef.current, newOptions, chartRef.current)) {
+      chartRef.current?.destroy();
       chartRef.current = undefined;
+      /**
+       * TODO: preserve zoom when new series is selected?
+       * may also want to preserve other user interactions with the chart
+       * by taking some things from chartRef.current and putting them in newOptions
+       */
+      newOptions = uPlot.assign(
+        optionsRef.current,
+        newOptions || {},
+        // { scales: scalesZoomData.current }, // not what we want
+      ) as uPlot.Options;
+      chartRef.current = new uPlot(newOptions, normalizedData as AlignedData, chartDivRef.current);
+    } else {
+      /**
+       * TODO: instead of returning true or false,
+       * shouldUpdate could return actions:
+       * 'noAction', 'resetSeries', and 'remount', etc.
+       * then the we could put addSeries and delSeries here
+       * or remount, etc.
+       * depending on the action returned
+       */
+      // chartRef.current?.redraw();
+    }
+    return () => {
+      if (options) optionsRef.current = newOptions as uPlot.Options;
```

Comment:
we are also using `optionsRef.current` in the initial chart mount and in `getNormalizedData`. We could replace it with prevOptions in those places but it would be confusing IMO. 

also overall, it would also take me some time to wrap my head around it enough to feel comfortable.